cloudfour/cloudfour.com-patterns
# Theo design tokens

## Overview

Resolves #471 by integrating Theo support into Sass stylesheets and Storybook stories. Both are accomplished via a small amount of custom code.

The result is that you can define properties in YAML files conforming to [the Theo spec](https://github.com/salesforce-ux/theo#spec), which you may then consume via Sass or JS. Changes made to these design tokens automatically refresh in Storybook.

Theo was chosen because its API was more flexible than Amazon's Style Dictionary, allowing direct integration with Sass and Webpack/Storybook. Other solutions would have required an intermediary step where files were generated and subsequently parsed.

The Sass portion is based on [Basalt's popular article on the subject](https://basalt.io/blog/theo-design-tokens-using-node-sass-importer-for-any-build-method), though simplified in one or two small areas.

I'd originally intended on using [Autodesk's theo-loader](https://github.com/Autodesk/theo-loader), but could not get it to work properly even in a new project. I'm guessing it may be out of date. It also had limitations as to the file types supported.

Luckily, writing a custom loader proved simpler than I anticipated.

## Testing

1. Check out branch locally.
1. `npm run storybook`
1. Edit the value of `blue` in `src/demo/design-tokens/colors.yml`. Observe changes applying to demo button and color page.
1. Edit the value of `brand_primary` directly in `src/demo/design-tokens/brand.yml`. Observe changes applying to demo button and color page.

## Screenshots

![storybook-theo](https://user-images.githubusercontent.com/69633/75181550-56b58600-56f3-11ea-97ab-ac925ed6afb9.gif)
